File: web/packages/core/src/labeling.ts
/**
 * Label assignment and offset-based replacement.
 *
 * Kept in behavioural parity with the Python core
 * (`src/prompt_anonymizer/labeling.py`): spans are merged score-first,
 * identical source strings of the same entity type share a label, and
 * replacement runs end-first so offsets stay valid.
 */

import type { EntitySpan, Language } from "./types.js";

export const LABELS: Record<Language, Record<string, string>> = {
  en: {
    PERSON: "Name",
    PERSON_FIRST_NAME: "First_Name",
    PERSON_MIDDLE_NAME: "Middle_Name",
    PERSON_LAST_NAME: "Last_Name",
    EMAIL_ADDRESS: "Email",
    LOCATION: "Location",
    PHONE_NUMBER: "Phone",
    JP_POSTAL_CODE: "PostalCode",
    JP_MY_NUMBER: "MyNumber",
    CREDIT_CARD: "CreditCard",
    US_SSN: "SSN",
    IBAN_CODE: "IBAN",
    CUSTOM: "Custom",
  },
  ja: {
    PERSON: "人名",
    PERSON_FIRST_NAME: "名",
    PERSON_MIDDLE_NAME: "ミドルネーム",
    PERSON_LAST_NAME: "姓",
    EMAIL_ADDRESS: "メールアドレス",
    LOCATION: "住所",
    PHONE_NUMBER: "電話番号",
    JP_POSTAL_CODE: "郵便番号",
    JP_MY_NUMBER: "マイナンバー",
    CREDIT_CARD: "クレジットカード",
    US_SSN: "社会保障番号",
    IBAN_CODE: "IBAN",
    CUSTOM: "秘匿情報",
  },
  es: {
    PERSON: "Nombre",
    PERSON_FIRST_NAME: "Nombre",
    PERSON_MIDDLE_NAME: "SegundoNombre",
    PERSON_LAST_NAME: "Apellido",
    EMAIL_ADDRESS: "Correo",
    LOCATION: "Dirección",
    PHONE_NUMBER: "Teléfono",
    JP_POSTAL_CODE: "CódigoPostal",
    JP_MY_NUMBER: "MyNumber",
    CREDIT_CARD: "Tarjeta",
    US_SSN: "SSN",
    IBAN_CODE: "IBAN",
    CUSTOM: "Personalizado",
  },
  vi: {
    PERSON: "Tên",
    PERSON_FIRST_NAME: "Tên",
    PERSON_MIDDLE_NAME: "TênĐệm",
    PERSON_LAST_NAME: "Họ",
    EMAIL_ADDRESS: "Email",
    LOCATION: "ĐịaChỉ",
    PHONE_NUMBER: "SốĐiệnThoại",
    JP_POSTAL_CODE: "MãBưuĐiện",
    JP_MY_NUMBER: "MyNumber",
    CREDIT_CARD: "ThẻTínDụng",
    US_SSN: "SSN",
    IBAN_CODE: "IBAN",
    CUSTOM: "TùyChỉnh",
  },
  zh: {
    PERSON: "姓名",
    PERSON_FIRST_NAME: "名",
    PERSON_MIDDLE_NAME: "中间名",
    PERSON_LAST_NAME: "姓",
    EMAIL_ADDRESS: "电子邮箱",
    LOCATION: "地址",
    PHONE_NUMBER: "电话号码",
    JP_POSTAL_CODE: "邮政编码",
    JP_MY_NUMBER: "MyNumber",
    CREDIT_CARD: "信用卡",
    US_SSN: "SSN",
    IBAN_CODE: "IBAN",
    CUSTOM: "自定义",
  },
  ko: {
    PERSON: "이름",
    PERSON_FIRST_NAME: "이름",
    PERSON_MIDDLE_NAME: "중간이름",
    PERSON_LAST_NAME: "성",
    EMAIL_ADDRESS: "이메일",
    LOCATION: "주소",
    PHONE_NUMBER: "전화번호",
    JP_POSTAL_CODE: "우편번호",
    JP_MY_NUMBER: "MyNumber",
    CREDIT_CARD: "신용카드",
    US_SSN: "SSN",
    IBAN_CODE: "IBAN",
    CUSTOM: "사용자지정",
  },
  fr: {
    PERSON: "Nom",
    PERSON_FIRST_NAME: "Prénom",
    PERSON_MIDDLE_NAME: "DeuxièmePrénom",
    PERSON_LAST_NAME: "NomDeFamille",
    EMAIL_ADDRESS: "Email",
    LOCATION: "Adresse",
    PHONE_NUMBER: "Téléphone",
    JP_POSTAL_CODE: "CodePostal",
    JP_MY_NUMBER: "MyNumber",
    CREDIT_CARD: "CarteBancaire",
    US_SSN: "SSN",
    IBAN_CODE: "IBAN",
    CUSTOM: "Personnalisé",
  },
  de: {
    PERSON: "Name",
    PERSON_FIRST_NAME: "Vorname",
    PERSON_MIDDLE_NAME: "Zweitname",
    PERSON_LAST_NAME: "Nachname",
    EMAIL_ADDRESS: "EMail",
    LOCATION: "Adresse",
    PHONE_NUMBER: "Telefon",
    JP_POSTAL_CODE: "Postleitzahl",
    JP_MY_NUMBER: "MyNumber",
    CREDIT_CARD: "Kreditkarte",
    US_SSN: "SSN",
    IBAN_CODE: "IBAN",
    CUSTOM: "Benutzerdefiniert",
  },
  pt: {
    PERSON: "Nome",
    PERSON_FIRST_NAME: "PrimeiroNome",
    PERSON_MIDDLE_NAME: "NomeDoMeio",
    PERSON_LAST_NAME: "Sobrenome",
    EMAIL_ADDRESS: "Email",
    LOCATION: "Endereço",
    PHONE_NUMBER: "Telefone",
    JP_POSTAL_CODE: "CódigoPostal",
    JP_MY_NUMBER: "MyNumber",
    CREDIT_CARD: "CartãoDeCrédito",
    US_SSN: "SSN",
    IBAN_CODE: "IBAN",
    CUSTOM: "Personalizado",
  },
  it: {
    PERSON: "Nome",
    PERSON_FIRST_NAME: "Nome",
    PERSON_MIDDLE_NAME: "SecondoNome",
    PERSON_LAST_NAME: "Cognome",
    EMAIL_ADDRESS: "Email",
    LOCATION: "Indirizzo",
    PHONE_NUMBER: "Telefono",
    JP_POSTAL_CODE: "CodicePostale",
    JP_MY_NUMBER: "MyNumber",
    CREDIT_CARD: "CartaDiCredito",
    US_SSN: "SSN",
    IBAN_CODE: "IBAN",
    CUSTOM: "Personalizzato",
  },
};

// Whitespace trimmed from the edges of remainder segments. An explicit set
// (rather than String.trim / str.strip) so both cores behave identically.
const STRIP_CHARS = " \t\n\r\u3000";

export type NamePart = "first" | "middle" | "last";

// Label keys for the localized part words (see LABELS above).
const NAME_PART_KEYS: Record<NamePart, string> = {
  first: "PERSON_FIRST_NAME",
  middle: "PERSON_MIDDLE_NAME",
  last: "PERSON_LAST_NAME",
};

// Surname particles for given-name-first languages: a token from this set
// (compared lowercase) starts the family name, so "Vincent van Gogh" splits
// into first="Vincent", last="van Gogh" rather than a bogus middle name.
const SURNAME_PARTICLES = new Set([
  "al",
  "bin",
  "binti",
  "da",
  "das",
  "de",
  "degli",
  "del",
  "della",
  "den",
  "der",
  "di",
  "dos",
  "du",
  "el",
  "la",
  "le",
  "ten",
  "ter",
  "van",
  "von",
]);

export interface NamePartSpan {
  part: NamePart;
  /** Offsets relative to the PERSON source string. */
  start: number;
  end: number;
}

/**
 * Split a PERSON source string into name parts, when possible.
 *
 * Returns parts in text order, or an empty array when the name has fewer
 * than two whitespace-separated tokens (e.g. "山田太郎" — splitting unspaced
 * CJK names would require a name dictionary, so they keep a plain person
 * label). Consecutive middle tokens form one contiguous part so each part
 * label maps to exactly one value.
 *
 * `familyNameFirst` follows the language's native order (ja/zh/ko/vi:
 * family name first); given-name-first languages additionally attach
 * surname particles ("van", "de", ...) to the last name.
 * Mirrors `split_person_name` in the Python core.
 */
export function splitPersonName(source: string, familyNameFirst: boolean): NamePartSpan[] {
  const tokens: Array<[number, number]> = [];
  let cursor = 0;
  while (cursor < source.length) {
    if (STRIP_CHARS.includes(source[cursor]!)) {
      cursor++;
      continue;
    }
    let end = cursor;
    while (end < source.length && !STRIP_CHARS.includes(source[end]!)) end++;
    tokens.push([cursor, end]);
    cursor = end;
  }
  if (tokens.length < 2) return [];

  if (familyNameFirst) {
    const parts: NamePartSpan[] = [{ part: "last", start: tokens[0]![0], end: tokens[0]![1] }];
    const middle = tokens.slice(1, -1);
    if (middle.length > 0) {
      parts.push({ part: "middle", start: middle[0]![0], end: middle[middle.length - 1]![1] });
    }
    parts.push({ part: "first", start: tokens[tokens.length - 1]![0], end: tokens[tokens.length - 1]![1] });
    return parts;
  }

  let lastStart = tokens.length - 1;
  for (let i = 1; i < tokens.length - 1; i++) {
    if (SURNAME_PARTICLES.has(source.slice(tokens[i]![0], tokens[i]![1]).toLowerCase())) {
      lastStart = i;
      break;
    }
  }
  const parts: NamePartSpan[] = [{ part: "first", start: tokens[0]![0], end: tokens[0]![1] }];
  const middle = tokens.slice(1, lastStart);
  if (middle.length > 0) {
    parts.push({ part: "middle", start: middle[0]![0], end: middle[middle.length - 1]![1] });
  }
  parts.push({ part: "last", start: tokens[lastStart]![0], end: tokens[tokens.length - 1]![1] });
  return parts;
}

/**
 * Resolve overlaps, keeping the higher score (longer span on ties).
 *
 * A span that overlaps an already-kept span is not dropped outright: the
 * parts not covered by kept spans survive as trimmed remainder spans.
 * Dropping the whole span would leak the non-overlapping text — e.g. an
 * NER address span that also covers an already-masked postal code.
 * When `text` is given, remainder edges are trimmed of whitespace.
 */
export function mergeSpans(spans: EntitySpan[], text?: string): EntitySpan[] {
  const ordered = [...spans].sort(
    (a, b) => b.score - a.score || b.end - b.start - (a.end - a.start) || a.start - b.start,
  );
  const kept: EntitySpan[] = [];
  for (const span of ordered) {
    const blockers = kept
      .filter((k) => k.start < span.end && span.start < k.end)
      .sort((a, b) => a.start - b.start);
    if (blockers.length === 0) {
      kept.push(span);
      continue;
    }
    const segments: Array<[number, number]> = [];
    let cursor = span.start;
    for (const blocker of blockers) {
      if (blocker.start > cursor) segments.push([cursor, blocker.start]);
      cursor = Math.max(cursor, blocker.end);
    }
    if (cursor < span.end) segments.push([cursor, span.end]);
    for (let [segStart, segEnd] of segments) {
      if (text !== undefined) {
        while (segStart < segEnd && STRIP_CHARS.includes(text[segStart]!)) segStart++;
        while (segEnd > segStart && STRIP_CHARS.includes(text[segEnd - 1]!)) segEnd--;
      }
      if (segEnd > segStart) {
        kept.push({ start: segStart, end: segEnd, entityType: span.entityType, score: span.score });
      }
    }
  }
  return kept.sort((a, b) => a.start - b.start);
}

export interface ApplyLabelsOptions {
  /**
   * Label name parts of multi-token PERSON spans individually
   * (`<Name_1_First_Name>` / `<人名_1_姓>` …), sharing one person index per
   * unique full name; a later single-token PERSON span matching an
   * already-seen part value reuses that part's label, so
   * "John Smith … John" stays consistent.
   */
  splitPersonNames?: boolean;
  /** Native name order (ja/zh/ko/vi: family name first). */
  familyNameFirst?: boolean;
}

export function applyLabels(
  text: string,
  spans: EntitySpan[],
  labels: Record<string, string>,
  options: ApplyLabelsOptions = {},
): { text: string; mapping: Record<string, string> } {
  const merged = mergeSpans(spans, text);

  const labelBySource = new Map<string, string>();
  const counters = new Map<string, number>();
  const mapping: Record<string, string> = {};
  const personIndexBySource = new Map<string, number>();
  const partLabelByValue = new Map<string, string>();
  const replacements: Array<{ start: number; end: number; label: string }> = [];

  for (const span of merged) {
    const source = text.slice(span.start, span.end);
    if (options.splitPersonNames && span.entityType === "PERSON") {
      const parts = splitPersonName(source, options.familyNameFirst ?? false);
      if (parts.length > 0) {
        if (!personIndexBySource.has(source)) {
          const next = (counters.get(span.entityType) ?? 0) + 1;
          counters.set(span.entityType, next);
          personIndexBySource.set(source, next);
        }
        const index = personIndexBySource.get(source)!;
        const personWord = labels[span.entityType] ?? span.entityType;
        for (const { part, start, end } of parts) {
          const value = source.slice(start, end);
          const partWord = labels[NAME_PART_KEYS[part]] ?? NAME_PART_KEYS[part];
          const label = `<${personWord}_${index}_${partWord}>`;
          mapping[label] ??= value;
          if (!partLabelByValue.has(value)) partLabelByValue.set(value, label);
          replacements.push({ start: span.start + start, end: span.start + end, label });
        }
        continue;
      }
      const reused = partLabelByValue.get(source);
      if (reused !== undefined) {
        replacements.push({ start: span.start, end: span.end, label: reused });
        continue;
      }
    }
    const key = `${span.entityType}\u0000${source}`;
    if (!labelBySource.has(key)) {
      const next = (counters.get(span.entityType) ?? 0) + 1;
      counters.set(span.entityType, next);
      const labelName = labels[span.entityType] ?? span.entityType;
      const label = `<${labelName}_${next}>`;
      labelBySource.set(key, label);
      mapping[label] = source;
    }
    replacements.push({ start: span.start, end: span.end, label: labelBySource.get(key)! });
  }

  let result = text;
  for (let i = replacements.length - 1; i >= 0; i--) {
    const { start, end, label } = replacements[i]!;
    result = result.slice(0, start) + label + result.slice(end);
  }

  return { text: result, mapping };
}

/** Restore original values; labels are replaced longest-first. */
export function deanonymize(text: string, mapping: Record<string, string>): string {
  const labels = Object.keys(mapping).sort((a, b) => b.length - a.length);
  let result = text;
  for (const label of labels) {
    result = result.split(label).join(mapping[label]!);
  }
  return result;
}

File: src/prompt_anonymizer/labeling.py
"""Label assignment and offset-based replacement.

This module is pure (no NLP engine required) so it can be unit-tested in
isolation and kept in behavioural parity with the TypeScript core in
``web/packages/core``.
"""

from __future__ import annotations

import re
from collections.abc import Sequence
from dataclasses import dataclass, field
from importlib import resources
from typing import Any

import yaml

LABEL_TEMPLATE = "<{label}_{index}>"

# Name-part labels append a localized part word to the person label, e.g.
# <Name_1_First_Name> / <人名_1_姓>. Parts of the same person share the
# person index, so an LLM can tell which parts belong together.
PART_LABEL_TEMPLATE = "<{label}_{index}_{part}>"

# Label-file keys for the localized part words (see labels/<lang>.yaml).
_NAME_PART_KEYS = {
    "first": "PERSON_FIRST_NAME",
    "middle": "PERSON_MIDDLE_NAME",
    "last": "PERSON_LAST_NAME",
}

# Surname particles for given-name-first languages: a token from this set
# (compared lowercase) starts the family name, so "Vincent van Gogh" splits
# into first="Vincent", last="van Gogh" rather than a bogus middle name.
_SURNAME_PARTICLES = frozenset(
    {
        "al",
        "bin",
        "binti",
        "da",
        "das",
        "de",
        "degli",
        "del",
        "della",
        "den",
        "der",
        "di",
        "dos",
        "du",
        "el",
        "la",
        "le",
        "ten",
        "ter",
        "van",
        "von",
    }
)


@dataclass(frozen=True)
class EntitySpan:
    """A detected PII span in the original text."""

    start: int
    end: int
    entity_type: str
    score: float


@dataclass
class AnonymizeResult:
    """Result of :meth:`PromptAnonymizer.anonymize`."""

    text: str
    mapping: dict[str, str]
    entities: list[EntitySpan] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return {
            "text": self.text,
            "mapping": self.mapping,
            "entities": [
                {
                    "start": e.start,
                    "end": e.end,
                    "entity_type": e.entity_type,
                    "score": e.score,
                }
                for e in self.entities
            ],
        }


def load_labels(language: str) -> dict[str, str]:
    """Load entity-type -> human label translations bundled with the package."""
    ref = resources.files("prompt_anonymizer").joinpath(f"labels/{language}.yaml")
    with ref.open("r", encoding="utf-8") as f:
        data = yaml.safe_load(f)
    if not isinstance(data, dict):  # pragma: no cover - corrupt package data
        raise ValueError(f"Invalid label file for language '{language}'")
    return {str(k): str(v) for k, v in data.items()}


def deny_list_spans(text: str, deny_list: Sequence[str]) -> list[EntitySpan]:
    """Substring search for deny-listed terms (labelled ``CUSTOM``).

    Presidio's deny_list uses ``\\b`` word boundaries, which never match
    between Japanese characters, so we match plain substrings instead.
    Mirrors ``detectDenyList`` in the TypeScript core.
    """
    spans: list[EntitySpan] = []
    for needle in deny_list:
        if not needle:
            continue
        start = text.find(needle)
        while start != -1:
            spans.append(
                EntitySpan(start=start, end=start + len(needle), entity_type="CUSTOM", score=1.0)
            )
            start = text.find(needle, start + len(needle))
    return spans


# NER entity types whose detected values are propagated to exact repeats.
_PROPAGATE_TYPES = frozenset({"PERSON", "LOCATION"})

# Word characters of space-delimited (Latin-script) text, used as a boundary
# guard so a propagated value never matches inside a longer word ("An" in
# "Anh"). CJK/kana neighbours are intentionally not word-like here so
# "山田太郎さん" still matches. Keep in parity with LATIN_WORD in the
# TypeScript core's recognizers.ts.
_LATIN_WORD = re.compile("[0-9A-Za-z\\u00C0-\\u024F\\u1E00-\\u1EFF]")


def propagate_entity_values(text: str, spans: Sequence[EntitySpan]) -> list[EntitySpan]:
    """Propagate NER-detected values to exact occurrences the model missed.

    NER models can skip a repeat of an entity they detected elsewhere in
    the same text (e.g. a person name right after an honorific at sentence
    start). An exact repeat of a detected value is the same PII, so every
    uncovered occurrence gets a span with the seed detection's type and
    score. Mirrors ``propagateEntityValues`` in the TypeScript core.
    """
    occupied = [(span.start, span.end) for span in spans]

    def overlaps(start: int, end: int) -> bool:
        return any(start < o_end and o_start < end for o_start, o_end in occupied)

    seeds: dict[tuple[str, str], float] = {}
    for span in sorted(spans, key=lambda s: s.start):
        if span.entity_type not in _PROPAGATE_TYPES:
            continue
        value = text[span.start : span.end]
        if len(value.strip()) < 2:
            continue
        key = (span.entity_type, value)
        seeds[key] = max(seeds.get(key, span.score), span.score)

    extra: list[EntitySpan] = []
    for (entity_type, value), score in seeds.items():
        start = text.find(value)
        while start != -1:
            end = start + len(value)
            at = start
            start = text.find(value, end)
            if overlaps(at, end):
                continue
            before = text[at - 1] if at > 0 else ""
            after = text[end] if end < len(text) else ""
            if (before and _LATIN_WORD.match(before) and _LATIN_WORD.match(value[0])) or (
                after and _LATIN_WORD.match(after) and _LATIN_WORD.match(value[-1])
            ):
                continue
            extra.append(EntitySpan(start=at, end=end, entity_type=entity_type, score=score))
            occupied.append((at, end))
    return extra


# Whitespace trimmed from the edges of remainder segments. An explicit set
# (rather than str.strip / String.trim) so both cores behave identically.
_STRIP_CHARS = " \t\n\r\u3000"


def merge_spans(spans: list[EntitySpan], text: str | None = None) -> list[EntitySpan]:
    """Resolve overlaps, keeping the higher score (longer span on ties).

    A span that overlaps an already-kept span is not dropped outright: the
    parts not covered by kept spans survive as trimmed remainder spans.
    Dropping the whole span would leak the non-overlapping text - e.g. an
    NER address span that also covers an already-masked postal code.
    When ``text`` is given, remainder edges are trimmed of whitespace.
    Returns spans sorted by start offset.
    """
    ordered = sorted(spans, key=lambda s: (-s.score, -(s.end - s.start), s.start))
    kept: list[EntitySpan] = []
    for span in ordered:
        blockers = sorted(
            (k for k in kept if k.start < span.end and span.start < k.end),
            key=lambda k: k.start,
        )
        if not blockers:
            kept.append(span)
            continue
        segments: list[tuple[int, int]] = []
        cursor = span.start
        for blocker in blockers:
            if blocker.start > cursor:
                segments.append((cursor, blocker.start))
            cursor = max(cursor, blocker.end)
        if cursor < span.end:
            segments.append((cursor, span.end))
        for seg_start, seg_end in segments:
            if text is not None:
                while seg_start < seg_end and text[seg_start] in _STRIP_CHARS:
                    seg_start += 1
                while seg_end > seg_start and text[seg_end - 1] in _STRIP_CHARS:
                    seg_end -= 1
            if seg_end > seg_start:
                kept.append(
                    EntitySpan(
                        start=seg_start,
                        end=seg_end,
                        entity_type=span.entity_type,
                        score=span.score,
                    )
                )
    return sorted(kept, key=lambda s: s.start)


def split_person_name(source: str, family_name_first: bool) -> list[tuple[str, int, int]]:
    """Split a PERSON source string into name parts, when possible.

    Returns ``(part, start, end)`` triples ("first" / "middle" / "last",
    offsets relative to ``source``) in text order, or an empty list when the
    name has fewer than two whitespace-separated tokens (e.g. "山田太郎" -
    splitting unspaced CJK names would require a name dictionary, so they
    keep a plain person label). Consecutive middle tokens form one
    contiguous part so each part label maps to exactly one value.

    ``family_name_first`` follows the language's native order (ja/zh/ko/vi:
    family name first); given-name-first languages additionally attach
    surname particles ("van", "de", ...) to the last name.
    Mirrors ``splitPersonName`` in the TypeScript core.
    """
    tokens: list[tuple[int, int]] = []
    cursor = 0
    while cursor < len(source):
        if source[cursor] in _STRIP_CHARS:
            cursor += 1
            continue
        end = cursor
        while end < len(source) and source[end] not in _STRIP_CHARS:
            end += 1
        tokens.append((cursor, end))
        cursor = end
    if len(tokens) < 2:
        return []

    if family_name_first:
        parts = [("last", tokens[0][0], tokens[0][1])]
        middle = tokens[1:-1]
        if middle:
            parts.append(("middle", middle[0][0], middle[-1][1]))
        parts.append(("first", tokens[-1][0], tokens[-1][1]))
        return parts

    last_start = len(tokens) - 1
    for i in range(1, len(tokens) - 1):
        if source[tokens[i][0] : tokens[i][1]].lower() in _SURNAME_PARTICLES:
            last_start = i
            break
    parts = [("first", tokens[0][0], tokens[0][1])]
    middle = tokens[1:last_start]
    if middle:
        parts.append(("middle", middle[0][0], middle[-1][1]))
    parts.append(("last", tokens[last_start][0], tokens[-1][1]))
    return parts


def apply_labels(
    text: str,
    spans: list[EntitySpan],
    labels: dict[str, str],
    *,
    split_person_names: bool = False,
    family_name_first: bool = False,
) -> tuple[str, dict[str, str]]:
    """Replace spans (end-first) with consistent labels.

    Identical source strings of the same entity type receive the same label.
    Returns the anonymized text and a ``label -> original`` mapping.

    With ``split_person_names``, multi-token PERSON spans are labelled per
    name part (``<Name_1_First_Name>`` / ``<人名_1_姓>`` ...), sharing one
    person index per unique full name; a later single-token PERSON span
    matching an already-seen part value reuses that part's label, so
    "John Smith ... John" stays consistent.
    """
    merged = merge_spans(spans, text)

    label_by_source: dict[tuple[str, str], str] = {}
    counters: dict[str, int] = {}
    mapping: dict[str, str] = {}
    person_index_by_source: dict[str, int] = {}
    part_label_by_value: dict[str, str] = {}
    replacements: list[tuple[int, int, str]] = []

    for span in merged:
        source = text[span.start : span.end]
        if split_person_names and span.entity_type == "PERSON":
            parts = split_person_name(source, family_name_first)
            if parts:
                if source not in person_index_by_source:
                    counters[span.entity_type] = counters.get(span.entity_type, 0) + 1
                    person_index_by_source[source] = counters[span.entity_type]
                index = person_index_by_source[source]
                person_word = labels.get(span.entity_type, span.entity_type)
                for part, rel_start, rel_end in parts:
                    value = source[rel_start:rel_end]
                    part_word = labels.get(_NAME_PART_KEYS[part], _NAME_PART_KEYS[part])
                    label = PART_LABEL_TEMPLATE.format(
                        label=person_word, index=index, part=part_word
                    )
                    mapping.setdefault(label, value)
                    part_label_by_value.setdefault(value, label)
                    replacements.append((span.start + rel_start, span.start + rel_end, label))
                continue
            if source in part_label_by_value:
                replacements.append((span.start, span.end, part_label_by_value[source]))
                continue
        key = (span.entity_type, source)
        if key not in label_by_source:
            counters[span.entity_type] = counters.get(span.entity_type, 0) + 1
            label_name = labels.get(span.entity_type, span.entity_type)
            label = LABEL_TEMPLATE.format(label=label_name, index=counters[span.entity_type])
            label_by_source[key] = label
            mapping[label] = source
        replacements.append((span.start, span.end, label_by_source[key]))

    result = text
    for start, end, label in reversed(replacements):
        result = result[:start] + label + result[end:]

    return result, mapping


def deanonymize(text: str, mapping: dict[str, str]) -> str:
    """Restore original values in ``text`` using an anonymize mapping.

    Labels are replaced longest-first to avoid partial-label collisions.
    """
    for label in sorted(mapping, key=len, reverse=True):
        text = text.replace(label, mapping[label])
    return text
